Allow sorting of group headers within any given hierarchy

awseward opened the pull request:
### Changes Made

Initially, this was implemented to order group headers within a given hierarchy according to the order they were encountered in (a decision made in order to respect the intent of [the original ListGroupable](https://github.com/procore/react-widgets/blob/bc933b37f1756979a784b6b806e6f3dc43670086/src/ListGroupable.jsx#L191-L193)).

This change allows us to sort the headings per hierarchy level.
### Example

Here's an example, in case that description above doesn't quite paint a clear picture.
##### Before

[The requirement for this](https://github.com/procore/wrench/pull/1283) was that the `Budgeted/Non-Budgeted` grouping needed to be sorted such that `Budgeted` always came first, but with the initial implementation of n-tier grouping, you're sort of just at the mercy of which groups were found first in the flat `data` array that the groups are constructed from.

This means that, depending on the data and the query filter, sometimes `Budgeted` would be first, but other times `Non-Budgeted` would be first. One such case of `Non-Budgeted` coming first is shown below.
<img width="510" alt="screen shot 2016-05-04 at 11 21 18" src="https://cloud.githubusercontent.com/assets/901035/15024610/660b9934-11ea-11e6-9125-b5851e2bf83a.png">
##### After

With this change, the caller can enforce an ordering for any given tier of group headers as shown below. (same data, same query filter). Notice how now the `Budgeted` group is shown before the `Non-Budgeted` group, as per the requirement.
<img width="502" alt="screen shot 2016-05-04 at 11 22 47" src="https://cloud.githubusercontent.com/assets/901035/15024656/942c5f88-11ea-11e6-8dea-7ab664543122.png">

awseward committed:
First stab at sorting different levels

awseward committed:
This'll do it

awseward committed:
Update built artifacts

awseward committed:
Fix compareIdentityFn

awseward committed:
Simplify _getPoppedArrayClone

awseward committed:
Use new groupBy form

Sort of a `groupDefinitions` kind of thing

awseward committed:
Update built artifacts

awseward committed:
Revert changes to PropTypes.accessor

awseward committed:
Update built artifacts

awseward committed:
Be safer about getting node children

awseward committed:
Update built artifacts

awseward committed:
Fix getcurrentNodeChildren

awseward committed:
Update built artifacts

remyyounes commented on src/util/objectTraversal.js:
```diff
@@ -24,7 +21,9 @@ export function depthFirst(currentNode, getChildren, onInternal, onLeaf, state)
   }
 
-  return getChildren(currentNode).reduce(
+  const getCurrentChildren = getChildren[0] || (x => Object.keys(s));
```
Is this supposed to be `(x => Object.keys(s))` or `(x => Object.keys(x))`

awseward replied:
@remyyounes: Ah, good catch.

pmkhougaz commented:
+1 pending Remy's comments

awseward committed:
Fix typo

awseward committed:
Update built artifacts

yoyis3000 commented:
+1

zgross commented:
+1

denzyl commented:
+1

kespinola commented:
+1
Sensible, but counter requirement. It feels like we need a groupByRules. Procore want more than just groupBy... it wants total list control.

awseward committed:
Pass in sort logic, not compare

This needs to be done, because giving `(a, b) => 0` to
Array.prototype.sort does not have the same "no sort" effect I thought
it would.

awseward committed:
Update built artifacts

thomasm-procore commented:
Passed testing.
